### PR TITLE
Define dedicated file configuration SPI ComponentProvider

### PR DESCRIPTION
**Pull request by jack-berg**
A while ago I opened PR #5912, which introduced ExtendedConfigProperties. The idea was to have an extended version of the ConfigProperties interface which included new accessors for reading structured data in addition to the simple types already exposed by ConfigProperties. Then, when file configuration needed to load a custom SDK extension component like SpanExporter, it would construct a ExtendedConfigProperties from the YAML contents and pass it to the SPI (i.e. ConfigurableSpanExporterProvider.

This approach is problematic because it requires SPI implementations to understand whether they're working on a file configuration context or not based on whether the ConfigProperties is actually an instance of ExtendedConfigProperties. The shape of data and the access patterns is quite different for file configuration vs env var / system property based configuration, so this whole thing is error prone and rather confusing.

In this PR, I propose a different direction which embraces the spec decision that file configuration should be thought of as the v2 of configuration, rather than something that is an extension of the existing env var / system property mechanism. Here's how I'm thinking about it:
- File configuration is a distinct configuration path from env var / system property configuration, including completely separate SPIs. 
  - If you set `otel.experimental.config.file` you're opting into this separate path. If you don't, all the existing SPIs and config mechanism works.
  - This separation is intention. It should be less confusing than the alternative where SPIs and ConfigProperties serve in multiple capacities.
- We introduce a generic ComponentProvider SPI, reflecting the [spec definition](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/file-configuration.md#component-provider).
  - ComponentProvider is a single interface to provide instances of all the SDK extension points (Sampler, SpanExporter, MetricExporter, Propagator, etc).
  - Users implement this SPI to provide custom named SDK extension components in a file configuration context. It is never called outside a file configuration context.
- We introduce a new StructuredConfigProperties for reading structured config data.
  - This interface is similar to ConfigProperties, but does not inherit from it.
  - It has new methods for reading structured data. Its interface for reading primitives is slightly modified to reflect the required semantics of representing a YAML node (i.e. getScalarList is nullable so callers can differentiate between not set and set but empty).
  - This interface is passed to `ComponentProvider#create(StructuredConfigProperties)`.
- Component authors who wish for their components to be accessible in both an env var / system property and an structured file config context must implement both SPIs. In doing so, they are made very aware of the differences between the two and difference in config schema.

For demonstration purposes and to make this PR as simple as possible, I've implemented a single ComponentProvider in the `opentelemetry-exporters-prometheus` artifact. If this PR is merged, I will followup by implementing ComponentProvider for the other built in components.

### CODE REVIEW

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/internal/PrometheusComponentProvider.java
Comment (jack-berg) [inline exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/internal/PrometheusComponentProvider.java:20]: A demonstration ComponentProvider. Notice how config properties are accessed without dot notation since the YAML node is presented directly as the `StructuredConfigProperties`.

File: sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/StructuredConfigProperties.java
Comment (jack-berg) [inline sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/StructuredConfigProperties.java:168]: `getStructured` and `getStructuredList` are the key bits which allow reading complex data.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdk.java
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdk.java:85]: This new method allows callers to determine if autoconfiguration was driven by file configuration or by env vars / system properties. Either this or `getConfig` will return null. The other will be non-null.

If file configuration is used, `getStructuredConfig` represents the full YAML config file.

File: sdk-extensions/autoconfigure/src/testFullConfig/java/io/opentelemetry/sdk/autoconfigure/FileConfigurationTest.java
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/testFullConfig/java/io/opentelemetry/sdk/autoconfigure/FileConfigurationTest.java:211]: This test demonstrates how a consumers like the otel java agent could access complex config data which is not part of the core file configuration schema. The `other` key and the structure it encloses is not part of the core schema yet its contents are accessible.

This will be important for configuring java agent specific options, like content for the JMX integration.

cc @trask, @laurit

File: sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/MetricReaderFactory.java
Comment (jack-berg) [inline sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/MetricReaderFactory.java:77]: Here's what it looks like to load a component in the file configuration implementation. We state that we want ` ComponentProvider` which provides instances of `MetricReader.class` named `prometheus`, and we pass the `prometheusModel` to it (which is translated to StructuredConfigProperties).

Nice and easy.

File: sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/ComponentProvider.java
Comment (breedx-splk) [inline sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/ComponentProvider.java:17]: This PR or a follow-up?

File: sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/StructuredConfigProperties.java
Comment (breedx-splk) [inline sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/StructuredConfigProperties.java:135]: Hmmm, do we foresee wanting lists of other non-string scalar types? I could see lists of integers, for example, being useful for defining histogram buckets or whatever. Probably plenty of other examples.  In any case, why not just call this `getStringList()`?

File: sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/YamlStructuredConfigProperties.java
Comment (breedx-splk) [inline sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/YamlStructuredConfigProperties.java:35]: There's a lot of work being done in this constructor! Might prefer to keep the constructor simple and just assign fields...which means making this a static factory method, something like `YamlStructuredConfigProperties.fromMap(...)`.

File: sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/YamlStructuredConfigProperties.java
Comment (breedx-splk) [inline sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/YamlStructuredConfigProperties.java:134]: 🙈

Comment (breedx-splk) [reply sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/YamlStructuredConfigProperties.java:134]: Probably worth a comment that calls out the loss of resolution if the value is actually a long.

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdk.java
Comment (breedx-splk) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdk.java:85]: This is kinda the only unfortunate downside of this approach that I can find thus far. It's a small drag that users of the autoconfigured sdk now have to be aware or otherwise check which config was used to get the data they need.

It feels like there is room to flatten the structured config into `ConfigProperties` and also to objectify the `ConfigProperties` into a structured thing...but we can have that conversation another time.

File: sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfigUtil.java
Comment (breedx-splk) [inline sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfigUtil.java:53]: Is this a leftover or a workaround for java's questionable type system? 🤔

**Review by breedx-splk — COMMENTED** [review]
@jack-berg this is a fantastic step forward on the config front!

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdk.java
Comment (jack-berg) [reply sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/AutoConfiguredOpenTelemetrySdk.java:85]: >  It's a small drag that users of the autoconfigured sdk now have to be aware or otherwise check which config was used to get the data they need.

I would characterize it slightly differently: Providers of custom components like exporters and samplers need to be aware that somebody can use file based config or flat system property / env var based config, and provide different SPI implementations for each. This is frustrating, but seems to be the inevitable result of otel getting as far as it did without a proper configuration story. As I mentioned in the PR description, I started with the opposite approach and tried to have a single unified `ConfigProperties` for both configuration routes. I've concluded while this is simpler from an API standpoint, its more confusing and brittle for users, since even though we can do some gymnastics to make `ConfigProperties` work for both cases, the data presented to SPI implementers will be different (i.e. the property names and structure and semantics will be different for flat env vars than in structured config). So there appears to be no getting around confronting the user with this difference. Might as well make it as explicit and obvious as possible.

File: sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/ComponentProvider.java
Comment (jack-berg) [reply sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/ComponentProvider.java:17]: Followup. As of now, this SPI is only used to load `PrometheusHttpServer`. Need to go through and add support for custom SDK extension components (SpanExporter, MetricExporter, Sampler, etc) one by one.

File: sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/StructuredConfigProperties.java
Comment (jack-berg) [reply sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/StructuredConfigProperties.java:135]: Yeah this is something I wasn't sure about. Thoughts:
- Adding type-specific accessors for all the different scalar types adds more surface API surface area. We want to support booleans, integers, floats, longs, and doubles for symmetry, with each having a "get or default" method. So 10 total methods. 
- The approach here is to have a single method, since everything has an equivalent string representation, and to rely on the caller to convert string to whatever scalar type they need to access. Less surface area, but more work for callers.

File: sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/YamlStructuredConfigProperties.java
Comment (jack-berg) [reply sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/YamlStructuredConfigProperties.java:134]: In a local version I got rid of the `getInt` versions of these methods for this reason. Figured that because the use case is config, its probably never going to be a problem to use the extra memory for the 64 bit long version.

I caved and restored the getInt accessors because sometimes you want to get a property like a port, and being forced to get a long and convert it down to an int to match the argument requirements of whatever library is being configured pushes the conversion problem to the caller.

I'll add a comment to the javadoc of getInt.

File: sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfigUtil.java
Comment (jack-berg) [reply sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfigUtil.java:53]: Needed to appease the type system:
```
/Users/jberg/code/open-telemetry/opentelemetry-java/sdk-extensions/incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/fileconfig/FileConfigUtil.java:54: error: incompatible types: cannot infer type-variable(s) R,A,CAP#1,T#2
            .collect(Collectors.toList());
                    ^
    (argument mismatch; Collector<ComponentProvider<?>,CAP#2,List<ComponentProvider<?>>> cannot be converted to Collector<? super ComponentProvider,CAP#2,List<ComponentProvider<?>>>)
```
We can't load a parameterized SPI type, so the prior line uses the raw type `List<ComponentProvider>` instead of `List<ComponentProvider<?>>` (casting isn't allowed). So we use the `.map` to coerce the list values from being raw types to having a parameter type of `?`. Its weird. I'm open to suggestions.

**Comment by jack-berg** [comment]
@breedx-splk I would greatly appreciate another look when you have a chance 🙂

File: sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/StructuredConfigProperties.java
Comment (breedx-splk) [reply sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/StructuredConfigProperties.java:141]: Seems fine to me. Thanks for talking it through.

**Review by breedx-splk — APPROVED** [review]
This looks like a good way to proceed. I wouldn't mind having another set of eyes from @trask or @laurit on the instrumentation side as well, but seems good from where I stand.

**Comment by jack-berg** [comment]
@open-telemetry/java-approvers will merge today tomorrow if there are no additional comments.

File: sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/ComponentProvider.java
Comment (trask) [inline sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/ComponentProvider.java:42]:
```suggestion
  // TODO consider dynamic configuration use case before stabilizing in case that affects any API decisions
  T create(StructuredConfigProperties config);
```

File: sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/StructuredConfigProperties.java
Comment (trask) [reply sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/StructuredConfigProperties.java:141]: unresolving to discuss a bit more

I think I kind of prefer being explicit about the types instead of converting to String and back (unless the configuration file format intentionally defines equivalency between number lists and equivalend string lists)

in this case, the underlying json schema data types are

* String
* Boolean
* Integer (mapped to Java's Long)
* Number (mapped to Java's Double)

we probably don't need the Java Integer and Float variants, similar to how we only support Long and Double attribute values

another possible option to keep the number of methods down could be something like `List<T> getList(String name, Class<T> type)` or `List<T> getScalarList(String name, Class<T> type)`

Comment (jack-berg) [reply sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/StructuredConfigProperties.java:141]: > List<T> getScalarList(String name, Class<T> type)

Let's do this and document that only certain known types are supported.

Comment (jack-berg) [reply sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/StructuredConfigProperties.java]: Pushed a commit to reflect this convo.

Comment (jack-berg) [reply sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/StructuredConfigProperties.java]: Meh I'm flip flopping. Dedicated methods for each type seems better now that I've written a single generic method:
- getDoubleList
- getLongList
- getStringList
- getBooleanList

**Pull request merged**
